marimo/_runtime/runtime.py

```diff
@@ -1088,40 +1092,38 @@ async def run(
                 filtered_requests.append(request)
 
         await self._run_cells(
-            self.mutate_graph(
-                filtered_requests,
-                deletion_requests=[
-                    # Always delete the scratchpad cell
-                    DeleteCellRequest(cell_id=SCRATCH_CELL_ID)
-                ],
-            )
+            self.mutate_graph(filtered_requests, deletion_requests=[])
         )
 
-    async def run_scratchpad(self, request: ExecuteScratchpadRequest) -> None:
-        LOGGER.warning("Running scratchpad cell")
+    async def run_scratchpad(self, code: str) -> None:
         roots = {SCRATCH_CELL_ID}
-        # Mutated graph by adding scratchpad cell
-        self.mutate_graph(
-            execution_requests=[
-                ExecutionRequest(SCRATCH_CELL_ID, request.code)
-            ],
-            deletion_requests=[],
-        )
+
+        # If cannot compile, don't run
+        try:
+            cell = compile_cell(code, SCRATCH_CELL_ID)
+            cell.defs.clear()  # remove definitions
+            cell.refs.clear()  # remove definitions
+        except SyntaxError:
+            return
+
+        # Create graph of just the scratchpad cell
+        graph = dataflow.DirectedGraph()
+        graph.register_cell(SCRATCH_CELL_ID, cell)
+
         runner = cell_runner.Runner(
             roots=roots,
-            graph=self.graph,
+            graph=graph,
             glbls=self.globals,
```

Comment:
@mscolnick — If you replace this with `copy.copy(self.globals)` (i.e. shallow copy), then the kernel globals won't get polluted. The scratch pad will still be able to mutate globals but that's fine, a deepcopy would be too expensive.

Reply:
I think it's preferable to be lax here, especially given that it's a scratchpad. The deepcopy will be too expensive in the common case of working with large dataframes, large arrays, etc, and it will also fail for non-pickleable objects.

Reply:
scratch won't be responsive, but that might be OK too. It could lead to other issues like an undefined ref (in the case a parent cell is compiled but hasn't run, like lazy) will throw a NameError, but strict mode would point out which cell has not run yet and prevent run.

An intermediate, but more verbose solution could be create `ScratchExecutor` that borrows some of the behavior from strict mode, but doesn't do the deep copying and can simplify the logic. More expensive than just a copy, but graph resolution for notebook sized graphs is still pretty cheap.

Then, for responsive runs, this function could register the cell here, only clear `defs`, and normally call `_run_cells(scratch)` (no need for another Runner). This would take maybe an override parameter

```python
Runner(...
   cell_execution_overrides : dict[Cell_Id: Execution_Type] = {SCRATCH : ScratchExecutor}
)
```

---

Anyway, just devils advocate, this is fine- and this comment can be revisited if scratch does cause reported issues

Reply:
Yea we can revisit. I dont think we want it to be responsive. and its intended to break outside the marimo reactive runtime. if you do want that, then just create a cell in notebook.